packages/block-editor/src/components/inserter/index.js

```diff
@@ -284,9 +328,37 @@ export default compose( [
 
 				const { insertBlock } = dispatch( blockEditorStore );
 
-				const blockToInsert = directInsertBlock?.length
-					? createBlock( ...directInsertBlock )
-					: createBlock( allowedBlockType.name );
+				let blockToInsert;
+
+				// Attempt to augment the directInsertBlock with attributes from an adjacent block.
+				// This ensures that styling from existing nearby blocks are preserved in the
+				// newly inserted block. To support intentionally clearing out certain attributes,
+				// the attributes of the directInsertBlock override those of the adjacent block.
+				// See: https://github.com/WordPress/gutenberg/issues/37904
+				if ( directInsertBlock ) {
+					const newAttributes = {};
+					const adjacentBlockAttributes = getAdjacentBlockAttributes();
+
+					Object.keys( adjacentBlockAttributes ).forEach(
+						( attributeName ) => {
+							if (
+								directInsertBlock.attributesToCopy?.[
+									attributeName
+								]
+							) {
+								newAttributes[ attributeName ] =
+									adjacentBlockAttributes[ attributeName ];
+							}
+						}
+					);
+
+					blockToInsert = createBlock( directInsertBlock.name, {
+						...newAttributes,
+						...( directInsertBlock.attributes || {} ),
```

Comment:
Just for my curiosity and to see if I missed a test case, did you have a use case in mind where the `directInsertBlock` would override attributes of the adjacent block?

Reply:
Good question. `directInsertBlock` could have attributes of its own defined; in case we define an attribute and also want to grab the same attribute from an adjacent block, it might make more sense for the adjacent one to override the initial.

Reply:
Agreed, thanks for asking the question! I wasn't _too_ sure of a use case, but thought it'd still be good to allow `attributes` to be defined in the `directInsertBlock`. Allowing the adjacent ones to override sounds good to me, I've swapped the order here 👍